### PR TITLE
GEOT-5844: Implemented the SumArea function for the Aggregate WPS pro…

### PR DESCRIPTION
**Pull request by mbarto**
…cess

### CODE REVIEW

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java [hunk 1]
```diff
@@ -907,7 +912,7 @@ public void encodeJoin(Type joinType, StringBuffer sql) {
     /**
      * Encodes the CREATE TABLE statement.
      * <p>
-     * Default implementation adds “CREATE TABLE” to the sql buffer.
+     * Default implementation adds â€œCREATE TABLEâ€� to the sql buffer.
```
Comment (bradh) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java]: Looks like some UTF-8 corruption.

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 2]
```diff
@@ -675,6 +683,17 @@ public void setAssociations(boolean foreignKeyGeometries) {
     }
     
     /**
+     * Returns the supported functions and the dialect name they map to.
+     */
+    public Map<String,String> getSupportedFunctions() {
+        if ( supportedFunctions == null ) {
+            supportedFunctions = new HashMap();
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Generics, new HashMap<>();

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 3]
```diff
@@ -1450,10 +1469,15 @@ protected Object getAggregateValue(FeatureVisitor visitor, SimpleFeatureType fea
             return null;
         }
         // try to extract an aggregate attribute from the visitor
-        AttributeDescriptor att = extractAggregateAttribute(visitor, featureType);
-        if (att == null && !isCountVisitor(visitor)) {
-            return null; // aggregate function optimization only supported for PropertyName expression
+        Expression aggregateExpression = getAggregateExpression(visitor);
+        // check if aggregation expression is supported
+        if (aggregateExpression instanceof FunctionExpression) {
+            FunctionExpression func = (FunctionExpression)aggregateExpression;
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Code conventions (space between cast and variable)

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 4]
```diff
@@ -1450,10 +1469,15 @@ protected Object getAggregateValue(FeatureVisitor visitor, SimpleFeatureType fea
...
+            if(!getSupportedFunctions().containsKey(func.getFunctionName().getName())) {
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Code conventions (space between if and opening parenthesis)

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 5]
```diff
@@ -3934,23 +3970,33 @@ protected void encodeAttribute(AttributeDescriptor attribute, Query query, Strin
...
         }
-        else {
+        else if (expression instanceof PropertyName){
             if ( att instanceof GeometryDescriptor ) {
                 encodeGeometryColumn((GeometryDescriptor)att, sql,query.getHints());
             }
             else {
                 dialect.encodeColumnName( att.getLocalName(), sql);
             }
+        } else if(expression instanceof FunctionExpression) {
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Coding conventions (same as above)

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java [hunk 6]
```diff
@@ -3934,23 +3970,33 @@ protected void encodeAttribute(AttributeDescriptor attribute, Query query, Strin
...
+            dialect.encodeFunction(getFunctioName(((FunctionExpression)expression).getName()), att.getLocalName(), sql);
+        } else {
+            sql.append( "*" );
         }
         
         sql.append(")");
     }
     
+    private String getFunctioName(String function) {
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: Confused, how do we reach here with a function name that is not supported? Wouldn't that be an error?

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java [hunk 7]
```diff
@@ -383,6 +384,10 @@ public void registerAggregateFunctions(Map<Class<? extends FeatureVisitor>,Strin
         aggregates.put( SumVisitor.class, "sum");
     }
     
+    public void registerFunctions(Map<String ,String> functions) {
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java]: Public method without javadoc

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java [hunk 8]
```diff
@@ -1463,4 +1468,18 @@ protected PrimaryKey getPrimaryKey(String typeName) throws IOException {
         SimpleFeatureType featureType = dataStore.getSchema(typeName);
         return dataStore.getPrimaryKey(featureType);
     }
+    
+    protected void encodeFunction( String function, String column, StringBuffer sql ) {
```
Comment (aaime) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java]: Functions in general have several parameters. See also https://github.com/geotools/geotools/blob/870e3ab8eece49c54186241e74adf6645b925f1c/modules/plugin/jdbc/jdbc-postgis/src/main/java/org/geotools/data/postgis/FilterToSqlHelper.java#L536 for existing support on encoding filter functions (not aggregate functions).

File: modules/library/main/src/main/java/org/geotools/feature/visitor/SumAreaVisitor.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ * 
+ *    (C) 2005-2017, Open Source Geospatial Foundation (OSGeo)
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/feature/visitor/SumAreaVisitor.java]: Just 2017

File: modules/library/main/src/main/java/org/geotools/feature/visitor/SumAreaVisitor.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+import org.geotools.factory.CommonFactoryFinder;
+import org.geotools.filter.IllegalFilterException;
+import org.opengis.feature.Feature;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.filter.FilterFactory;
+import org.opengis.filter.expression.Expression;
+
+public class SumAreaVisitor extends SumVisitor {
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/feature/visitor/SumAreaVisitor.java]: New public class without javadocs

File: modules/library/main/src/test/java/org/geotools/feature/visitor/SumAreaVisitorTest.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import org.opengis.feature.simple.SimpleFeatureType;
+
+import com.vividsolutions.jts.geom.Coordinate;
+import com.vividsolutions.jts.geom.GeometryFactory;
+import com.vividsolutions.jts.geom.LinearRing;
+import com.vividsolutions.jts.geom.Polygon;
+
+
+@RunWith(Parameterized.class)
+public class SumAreaVisitorTest<T> extends VisitorTestCase<T, T> {
```
Comment (aaime) [inline modules/library/main/src/test/java/org/geotools/feature/visitor/SumAreaVisitorTest.java]: Reformat file, several formatting issues (can be done fully since it's new)

File: modules/library/main/src/test/java/org/geotools/feature/visitor/SumAreaVisitorTest.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    private static final GeometryFactory GF = new GeometryFactory();
+    
+    public SumAreaVisitorTest(Class<T> valueClass, List<T> values, T expectedValue) {
+        super(valueClass, values, expectedValue);
+    }
+
+    @Parameterized.Parameters
+    public static Collection<Object[]> parameters() {
+
+        return Collections.singleton(new Object[]{Polygon.class, Arrays.asList(polygon( new int[]{ 12,6, 14,8, 16,6, 16,4, 12,4, 12,6} )), 12.0});
```
Comment (aaime) [inline modules/library/main/src/test/java/org/geotools/feature/visitor/SumAreaVisitorTest.java]: Just one test with one polygon, it's not checking for actual aggregation. I'd add a few more tests here, empty set, more than one, maybe see what happens with invalid polygons.

File: modules/library/main/src/main/java/org/geotools/feature/visitor/SumAreaVisitor.java (hunk 9 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/feature/visitor/SumAreaVisitor.java]: Btw, did not check the other new files, but in case, it would be the same

**Comment by aaime** [comment]
Seeing lots of test and implementation of aggregation in several databases, that's good, which ones have you actually tested? Are the paths of native aggregation and in memory aggregation both tested?

**Comment by mbarto** [comment]
@aaime, I tested (run online tests) all 3 that have native support at the moment (PostGIS, SQLServer, Oracle). I also tested H2 that has no native support. The other ones should behave as H2.

testSumArea automatically checks for native support and tests if in memory or native aggregation happened (looking at the visited flag).

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java (hunk 6 above)
Comment (mbarto) [reply modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java]: You are right, it cannot happen, I will refactor the method

**Comment by tbarsballe** [comment]
This looks ready for merge?

@aaime - any other feedback, or are you happy with the changes?

File: modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java [hunk 13]
```diff
@@ -384,6 +385,16 @@ public void registerAggregateFunctions(Map<Class<? extends FeatureVisitor>,Strin
     }
     
     /**
+     * Register the dialect mappings between Geotools FilterFunction names and database
+     * related function.
+     * 
+     * @param functions mappings from GT FilterFunction name to sql function name
+     */
+    public void registerFunctions(Map<String ,String> functions) {
```
Comment (jodygarnett) [inline modules/library/jdbc/src/main/java/org/geotools/jdbc/SQLDialect.java]: This is a new public method for SQLDialect opening up an ability to extend for client applications; do we need a proposal here to consider this API change?
